From 4040413e3ed35e1e43e033fc8017e94ea968a6da Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Mon, 14 Aug 2023 17:17:05 +0800
Subject: [PATCH 0283/1306] winex11.drv: Set _NET_WM_FULLSCREEN_MONITORS only
 when necessary.

If _NET_WM_FULLSCREEN_MONITORS is set then the property needs to be updated because it can't
be deleted by sending a _NET_WM_FULLSCREEN_MONITORS client message to the root window
according to the wm-spec version 1.4. Having the window spanning more than two monitors also
needs the property set. In other cases, _NET_WM_FULLSCREEN_MONITORS doesn't need to be set.
What's more, setting _NET_WM_FULLSCREEN_MONITORS adds a constraint on Mutter so that such a
window can't be moved to another monitor by using the Shift+Super+Up/Down/Left/Right
shortcut. So the property should be added only when necessary.

CW-Bug-Id: #22540
---
 dlls/winex11.drv/window.c | 32 ++++++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index d1ca193fa1f..c952a40f623 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -1244,6 +1244,38 @@ static void update_net_wm_fullscreen_monitors( struct x11drv_win_data *data )
         && !data->net_wm_fullscreen_monitors_set)
         return;
 
+    /* If _NET_WM_FULLSCREEN_MONITORS is not set and the fullscreen monitors are spanning only one
+     * monitor then do not set _NET_WM_FULLSCREEN_MONITORS.
+     *
+     * If _NET_WM_FULLSCREEN_MONITORS is set then the property needs to be updated because it can't
+     * be deleted by sending a _NET_WM_FULLSCREEN_MONITORS client message to the root window
+     * according to the wm-spec version 1.4. Having the window spanning more than two monitors also
+     * needs the property set. In other cases, _NET_WM_FULLSCREEN_MONITORS doesn't need to be set.
+     * What's more, setting _NET_WM_FULLSCREEN_MONITORS adds a constraint on Mutter so that such a
+     * window can't be moved to another monitor by using the Shift+Super+Up/Down/Left/Right
+     * shortcut. So the property should be added only when necessary. */
+    if (monitors[0] == monitors[1] && monitors[1] == monitors[2] && monitors[2] == monitors[3])
+    {
+        unsigned long count, remaining;
+        BOOL prop_found = FALSE;
+        long *prop_data;
+        int format;
+        Atom type;
+
+        if (!XGetWindowProperty( data->display, data->whole_window,
+                                 x11drv_atom(_NET_WM_FULLSCREEN_MONITORS), 0, ~0, False,
+                                 XA_CARDINAL, &type, &format, &count, &remaining,
+                                 (unsigned char **)&prop_data ))
+        {
+            if (type == XA_CARDINAL && format == 32 && count == 4)
+                prop_found = TRUE;
+            XFree(prop_data);
+        }
+
+        if (!prop_found)
+            return;
+    }
+
     if (!data->mapped)
     {
         XChangeProperty( data->display, data->whole_window, x11drv_atom(_NET_WM_FULLSCREEN_MONITORS),
-- 
2.43.2

